Performance over time benchmark blog post

Alex-Monahan opened the pull request:
Hi folks! Here is my first version of the DuckDB performance over time blog post! I welcome your feedback!! Goal is to publish this Friday, June 28th.

Total performance is up 14x in 3 years... Pretty wild! Individual benchmarks range from 3x-25x. 

I used the H2O.ai benchmark to measure both performance and scale. I also added a totally new window function benchmark with 16 different queries to really exercise the range of window functionality (get it? :-) ). I also tested import and export to Pandas, Arrow, and Parquet.

On the blogging side, I introduced Plotly.js so that we can have interactive visuals since there are so many results to display. Each chart has its own JSON config exported from the Python Plotly script I used to produce the result, so it is a little bit verbose. I do not expect any manual adjustments of those JSON files!

I have many formatting tasks yet to complete - I am happy to take a first pass at that after I see how the CI/CD treats me...

szarnyasg commented:
Hi Alex,

Great work. I have one major comment. The "good stuff", i.e. first plot comes ~1500 words into the article. I'm a bit concerned many readers will not make it is this far. Especially for ones who are unfamiliar with the H2O.ai benchmark, the first two blocks are difficult/uninteresting to read. To give concrete examples on what could be shortened:

1. The section "Python Client" is not necessary

    > To measure interoperability with other dataframe formats, we have used Python rather than R (used by H2O.ai) for this analysis. We do continue to use R for the data generation step for consistency with the benchmark. Python is DuckDB's most popular client, so this is also the most representative of real world performance.

    This is redundant as the summary above already says that we use "* Python instead of R"

2. There are a lot of technical details in the section "Export and Replacement Scans". Many of the readers don't know what a replacement scan is. It's important to discuss these differences but maybe this is best done in the appendix.

Alex-Monahan commented:
Great feedback! I will make some adjustments. Maybe we could even just have the first results chart be almost the first thing?

szarnyasg commented:
Yes, that sounds great!

One thing that we could link is the release calendar: https://duckdb.org/docs/dev/release_calendar

Alex-Monahan commented on _posts/2024-04-10-benchmarks-over-time.md:
```diff
@@ -0,0 +1,547 @@
+### DuckDB Versions
+
+Version 0.2.7 was the first version to include a Python client compiled for ARM64, so it was the first version that could easily run on the benchmarking compute resources. 
+Version 1.0.0 is the latest available at the time of publication, although we also provide a sneak preview of an in-development feature branch.
```
At the time I had run it, I was pointed at feature! Things move pretty fast around here...haha. 

It would be maybe an hour of work to re-run the analysis on main and re-output all the charts if that would be better. (it would take ~5 hours of runtime if we redo the scale one, so that would have to wait to run overnight).

Alex-Monahan replied:
Keeping feature due to timing!

Alex-Monahan commented on _posts/2024-04-10-benchmarks-over-time.md:
```diff
@@ -0,0 +1,547 @@
+### Default Settings
+
+All versions were run with the default settings. 
+The only small exception to this was specifying a temporary directory to enable intermediate calculations to spill to disk.
```
Then I set that setting when I didn't need to! I will omit the mention of it, since I effectively just changed the directory rather than set it.

Alex-Monahan replied:
Thanks!

Alex-Monahan commented on _posts/2024-04-10-benchmarks-over-time.md:
```diff
@@ -0,0 +1,547 @@
+
+Group by or aggregation operations are a critical OLAP workload, and have therefore received substantial focus in DuckDB, **improving over 12x in the last 3 years**.
+
+In version 0.3.1, multithreaded aggregation was enabled by default, providing a significant speedup.
```
I can do that, sure! I had to make a mapping of it for the script anyway. The first draft had used a by-version chart instead of by time, but I should have updated the language!

Alex-Monahan replied:
Made these changes, thank you!

Alex-Monahan commented on _posts/2024-04-10-benchmarks-over-time.md:
```diff
@@ -0,0 +1,547 @@
+
+#### Exporting Apache Arrow vs. Pandas vs. Parquet
+
+<div id="perf_over_time_export_arrow_pandas_parquet" style="width:100%;height:400px;"></div>
```
In all of the plots except for the specific Arrow/Pandas/Parquet plots, I have a cutoff point (version 0.5.1) where before that, I use Pandas and afterwards I use Arrow. That was to try and mirror the real world where folks probably wouldn't move over to Arrow until it was better, and partially to avoid having that big spike in the summary graphs. 

So the purpose of that plot is to include that big spike and show when Arrow starts to beat Pandas. 

That sound ok, or still not something you want included?

Alex-Monahan replied:
oh, looking more closely, I think that may just be a mistake on my side. Let me double check.

szarnyasg replied:
I tried to take the cutoff into consideration but even that didn't explain it.

Alex-Monahan replied:
Ok, I double checked, and the differences are:
* That 0.5.1 Pandas/Arrow cutoff 
* Area vs. line like you mentioned
* The line plot uses seconds instead of a relative measure, so that it is easier to add/remove series on

I would like to keep it in if possible since I think it tells an interesting story. Are there changes you would like me to make to clarify things a bit?

Also, totally agree on the placeholder - I will fix that now!

szarnyasg replied:
I still do not understand the two plots in my [original comment](https://github.com/duckdb/duckdb-web/pull/3146#discussion_r1650371546), and in particular I don't understand how the cutoff works. If we change from Pandas to Arrow at version 0.5.1, than the starting point of the area "Export group by results to Arrow" in the top diagram should be 0.5.1. But it stretches across the entire version timeline, including versions prior to 0.5.1.

Alex-Monahan replied:
Placeholders fixed on those 2 Pandas/Arrow/Parquet plots!

Alex-Monahan replied:
Gabor had some great comments, so I've added placeholder zeroes in the area charts (so that when you filter on just 1 series, you can see the correct date axis), and I added in some explanations for the line charts to explain why they are different. Thanks!

Alex-Monahan commented on _posts/2024-04-10-benchmarks-over-time.md:
```diff
@@ -0,0 +1,547 @@
+This allows for DuckDB to aggregate 1 Billion rows of data (50GB in size) on a Macbook Pro with only 16GB of RAM, even when the number of unique groups in the group by is large. 
+This represents at least a 10x improvement in aggregate processing scale over the course of the 3 years of the benchmark.
+
+DuckDB's hash join operator has supported larger than memory joins since version 0.6.1. 
```
I think it may have been the progress bar causing a huge log file in some way. I didn't investigate further though!

szarnyasg replied:
Thanks. This is not too important so we can drop this comment if you want to.

carlopi commented on _posts/2024-04-10-benchmarks-over-time.md:
```diff
@@ -0,0 +1,548 @@
+### H2O.ai as the Foundation
+
+This post measures DuckDB's performance over time on the H2O.ai benchmark for both joins and group by queries.
+Please see our previous [blog](https://duckdb.org/2023/04/14/h2oai.html) [posts](https://duckdb.org/2023/11/03/db-benchmark-update.html) for details on why we believe the H2O.ai benchmark is a good approach!
```
H20.ai has been updated recently, I will also put the stable website somewhere.

Alex-Monahan replied:
Oh, thank you!

Alex-Monahan replied:
I added the link you sent - thanks!

carlopi commented on _posts/2024-04-10-benchmarks-over-time.md:
```diff
@@ -0,0 +1,548 @@
+
+To measure interoperability with other dataframe formats, we have used Python rather than R (used by H2O.ai) for this analysis.
+We do continue to use R for the data generation step for consistency with the benchmark.
+Python is DuckDB's most popular client, so this is also the most representative of real world performance.
```
```suggestion
Python is DuckDB's most popular client, and it's my preferred tool to hack things together.```
Or equivalent, I would not put down an objective reason (that looks a bit weak to me) but go with something that says that writing Python is easier on you.

Alex-Monahan replied:
Fair point! I can make that change - it was a significant factor... :-)

Alex-Monahan replied:
Adjusted! Thanks!

carlopi commented on _posts/2024-04-10-benchmarks-over-time.md:
```diff
@@ -0,0 +1,548 @@
+### Workload Size
+
+We test only the middle 5GB dataset size for the workloads mentioned thus far, primarily because some import and export operations to external formats like Pandas must fit in memory (and we used a MacBook Pro M1 with only 16GB of RAM). 
+Additionally, running the tests for 3 year's of DuckDB versions was time-intensive even at that scale, due to the performance of older versions.
```
```suggestion
Additionally, running the tests for X different DuckDB version was time-intensive even at that scale, due to the performance of older versions.
```
I would put an explicit number on how many versions in total.

Alex-Monahan replied:
Happy to do that!

Alex-Monahan replied:
Made the change, thanks!

Alex-Monahan commented on _posts/2024-04-10-benchmarks-over-time.md:
```diff
@@ -0,0 +1,547 @@
+More recently, the DuckDB Labs team focused on scalability by developing algorithms that support larger than memory calculations. 
+We will see the fruits of those labors in the scale section later on!
+In addition, DuckDB focused exclusively on bug fixes in versions 0.10.1, 0.10.2, and 0.10.3 in preparation for an especially robust DuckDB 1.0. 
+The pre-0.9 trend is a better indicator for the future now that those two major milestones were accomplished!
```
Yes, those were the two! I can clarify that. 

I also agree that multi-threading only happens once, so I will reword that a bit.

Alex-Monahan replied:
Feel free to reopen this or reword things, but it is better than before at least!

carlopi commented on _posts/2024-04-10-benchmarks-over-time.md:
```diff
@@ -0,0 +1,548 @@
+* DuckDB Versions 0.2.7 through 1.0.0
+    * Nearly 3 years, from 2021-06-14 to 2024-06-03
+* Default settings
+
```
Should there be somewhere some pseudocode like:


```python
for v in duckdb_versions:
     pip install duckdb --version=v
     time python3 mytest.py
```
or whatever equivalent?

Alex-Monahan replied:
Well, unfortunately it is a fair bit uglier than that! I had to diagnose which versions of Pandas and Arrow would be compatible (using release dates and some trial and error). So it's 200-300 lines of Python to loop through all the versions and kick things off, and an 800 line Python script to run all of the benchmarks.

Alex-Monahan replied:
I'm going to close this for now, although we want to eventually share the code. Here is the repo I have been using for my analysis - it needs to be cleaned up a bit (some paths are hardcoded, etc.). I made it public, but I don't think I want to link to it in the blog.

https://github.com/Alex-Monahan/duckdb_benchmarks_over_time

carlopi commented on _posts/2024-04-10-benchmarks-over-time.md:
```diff
@@ -0,0 +1,548 @@
+
+If you have made it this far, welcome to the flock! 🦆
+[Join us on Discord](https://discord.duckdb.org/), we value your feedback!
+
```
Can the data be exported as a duckdb database file, and saved somewhere?

So people can run their own queries on it?
We have a blobs.duckdb.org to host random datasets, this can easily be one of the

This implies some sort of commitment to update it, so maybe do save the scripts used in some repo somewhere.

Alex-Monahan replied:
Ah, we can do that! I can provide the output as both a SQLite file and a DuckDB file (I used SQLite so that I could check on the progress of the benchmark while it was running in a separate process). Right now I have them in a few separate files, especially for the scale one since it kept running my laptop out of hard disk space... ;-). I will consolidate them.

How do I upload to blobs.duckdb.org?

I should save the scripts - good point. The results analysis are in a pair of Jupyter notebooks and they ended up getting pretty messy with all the combinations of visualizations that I tried. The script for running the benchmark is better, but I'd love a code review on that as well (but maybe should add some comments first...).

carlopi reviewed (commented):
Great, some comments here and there, but super interesting!

Alex-Monahan commented:
> Hi Alex,
> 
> Great work. I have one major comment. The "good stuff", i.e., the "Overall Benchmark Results" section and the first plot come ~1500 words into the article. I'm a bit concerned many readers will not make it is this far. Especially for ones who are unfamiliar with the H2O.ai benchmark, the first two blocks are difficult/uninteresting to read. To give concrete examples on what could be shortened:
> 
> 1. The section "Python Client" is not necessary
>    > To measure interoperability with other dataframe formats, we have used Python rather than R (used by H2O.ai) for this analysis. We do continue to use R for the data generation step for consistency with the benchmark. Python is DuckDB's most popular client, so this is also the most representative of real world performance.
>    
>    
>    This is redundant as the summary above already says that we use "* Python instead of R"
> 2. There are a lot of technical details in the section "Export and Replacement Scans". Many of the readers don't know what a replacement scan is. It's important to discuss these differences but maybe this is best done in the appendix.

I made some adjustments and I am really glad you mentioned this! 

I added the performance over time chart at the top of the post. 

I moved the benchmark design to the appendix, except for the summary which I left inline. 

Let me know what you think!

szarnyasg commented:
@Alex-Monahan the blog post currently starts with "In the last 3 years, DuckDB has become 3-25x faster and can analyze ~10x larger datasets."

I think it would be worth adding that this is **on the same hardware**.

We can remark that memory gotten 60% cheaper (!!) over the 3-year period between 2020 and 2023 (so it's shifted by one year): https://ourworldindata.org/grapher/historical-cost-of-computer-memory-and-storage?yScale=linear&time=2020..latest&facet=none

So if you spend the same amount on RAM, you can buy more than 2x amount the memory, and roughly deal with 2x the data. (It's not fully linear but it's not "very superlinear" either.)

Alex-Monahan commented on _posts/2024-04-10-benchmarks-over-time.md:
```diff
@@ -0,0 +1,547 @@
+The result of each H2O.ai query is written to a table in a persistent DuckDB file.
+This does require additional work when compared with an in-memory workflow (especially the burden on the SSD rather than RAM), but improves scalability and is a common approach for larger analyses.
+
+As in the current H2O.ai benchmark, categorical-type columns (`VARCHAR` columns with low cardinality) were converted to the `ENUM` type as a part of the benchmark.
```
Made the change by moving it to the appendix!

Alex-Monahan commented on _posts/2024-04-10-benchmarks-over-time.md:
```diff
@@ -0,0 +1,547 @@
+
+### Export and Replacement Scans
+
+However, we now extend this benchmark in several important ways. 
```
Fixed, thank you!

Alex-Monahan commented on _posts/2024-04-10-benchmarks-over-time.md:
```diff
@@ -0,0 +1,547 @@
+
+### Scale Tests
+
+However, using only 5GB of data does not answer our second key question: "What scale of data can it handle?"!
```
Fixed, thank you!

Alex-Monahan commented:
Also, I really like your changes to the intro - thank you!

Alex-Monahan commented:
Ok, I added a section to show RAM and SSD prices, inside the Scale tests section! I also renamed the file to the correct date. 

I feel like we are just about ready to go! Only remaining bit is to upload the raw dataset (I need to consolidate a few files together) and mention that in the post. 

This is much better thanks to the great feedback and fixes from @szarnyasg and @carlopi - thank you!

Alex-Monahan commented:
Ah, realized I need to change the link formatting. I will do that this evening as well!

Alex-Monahan commented:
Alrighty, all set once we get the results DB uploaded and a DuckDB web shell custom URL built and linked in the post!

When do we want to publish? MotherDuck is happy to re-share!

szarnyasg commented on _posts/2024-06-26-benchmarks-over-time.md:
```diff
@@ -0,0 +1,608 @@
+    (PARTITION BY id2 ORDER BY id3) AS first_by_id2_ordered_by_id3
+
+/* 307 Lead and Lag */
+first(v2) OVER
```
are these window function calls `first(v2)` instead of `lag(v2)` and `lead(v2)?

Alex-Monahan replied:
Yes, just because I like to use the generic window function syntax since I have it memorized by now. Functionally the same though, at least that is the intention!

Alex-Monahan commented:
Ok, we are good to go! Thanks again folks!
